langkit/templates/pkg_implementation_spec_ada.mako

```diff
@@ -558,6 +560,260 @@ private package ${ada_lib_name}.Implementation is
+
+   type Env_Name_Internal_Array is array (Positive range <>) of Symbol_Type;
+   type Env_Name_Record (Size : Positive) is record
+      Symbols : Env_Name_Internal_Array (1 .. Size);
+
+      Hash : Hash_Type;
+      --  Precomputed hash for this name. No need to compute the hash more than
+      --  once.
+   end record;
+
+   procedure Destroy is new Ada.Unchecked_Deallocation
+     (Env_Name_Record, Env_Name);
+   function Equivalent (Left, Right : Env_Name) return Boolean
+   is (Left.Symbols = Right.Symbols);
+   function Hash (Name : Env_Name) return Hash_Type is (Name.Hash);
+
+   package Env_Name_Sets is new Ada.Containers.Hashed_Sets
+     (Element_Type        => Env_Name,
+      Hash                => Hash,
+      Equivalent_Elements => Equivalent);
+   --  Internalization table for named symbols (see the Create_Env_Name
+   --  function below).
+
+   --  Tables to populate lexical entries in named envs
+
+   package NED_Assoc_Maps is new Ada.Containers.Hashed_Maps
+     (Key_Type        => Symbol_Type,
+      Element_Type    => Internal_Map_Node_Vectors.Vector,
+      Hash            => Hash,
+      Equivalent_Keys => "=",
+      "="             => Internal_Map_Node_Vectors."=");
+   --  Symbol/lexical env entry mappings for a given named env descriptor.
+   --  Symbols are not unique in all mappings, so the lexical env entries are
+   --  stored in a vector.
+
+   procedure Add
+     (Self : in out NED_Assoc_Maps.Map;
+      Key  : Symbol_Type;
+      Node : AST_Envs.Internal_Map_Node);
+   --  Add a symbol/lexical env entry mapping in Self
+
+   procedure Remove
+     (Self : in out NED_Assoc_Maps.Map;
+      Key  : Symbol_Type;
+      Node : ${T.root_node.name});
+   --  Remove a symbol/lexical env entry mapping from Self
+
+   --  Global table for named environments
+
+   package Sorted_Env_Maps is new Ada.Containers.Ordered_Maps
+     (Key_Type     => ${T.root_node.name},
+      Element_Type => Lexical_Env);
+   --  List of lexical environments, sorted by owning node. This means that the
+   --  following must be true for all cursors in such maps::
+   --
+   --     Key (Cur) = Element (Cur).Env.Node
+
+   package Node_Sets is new Ada.Containers.Hashed_Sets
+     (Element_Type        => ${T.root_node.name},
+      Hash                => Hash,
+      Equivalent_Elements => "=");
+
+   type Named_Env_Descriptor is record
+      Name : Env_Name;
+      --  Name corresponding to this descriptor. Useful during debugging.
+
+      Envs : Sorted_Env_Maps.Map;
+      --  For each env name, we can have one or several environments
+      --  (concurrent definitions). Just like foreign nodes in lexical
+      --  environments, we keep them sorted by node to preserve determinism:
+      --  given a set of loaded units, we will always have the same set of
+      --  name:env associations sorted in the same order and thus always the
+      --  same results at lookup time.
+
+      Env_With_Precedence : Lexical_Env;
```

Comment:
It’s indeed useful for brevity, but the reason for this is that it can temporarily become `Empty_Env` during invalidation (registration of new env that will take precedence, before we performed relocations). I forgot to describe this case: I’ll amend the comment.

Reply:
SGTM!